prow/plugins/milestoneapplier/milestoneapplier.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package milestoneapplier implements the plugin to automatically apply
+// the configured milestone after a PR is merged.
+package milestoneapplier
+
+import (
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/plugins"
+	"k8s.io/test-infra/prow/plugins/milestone"
+)
+
+const pluginName = "milestoneapplier"
+
+type githubClient interface {
+	SetMilestone(org, repo string, issueNum, milestoneNum int) error
+	ListMilestones(org, repo string) ([]github.Milestone, error)
+}
+
+func init() {
+	plugins.RegisterPullRequestHandler(pluginName, handlePullRequest, helpProvider)
+}
+
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	configInfo := map[string]string{}
+	for _, orgRepo := range enabledRepos {
+		configInfo[orgRepo] = fmt.Sprintf("The configured branches and milestones for this repo are: %v", config.MilestoneApplier[orgRepo].BranchToMilestone)
+	}
+
+	// The {WhoCanUse, Usage, Examples} fields are omitted because this plugin is not triggered with commands.
+	return &pluginhelp.PluginHelp{
+		Description: "The milestoneapplier plugin automatically applies the configured milestone for the base branch after a PR is merged. If a PR targets a non-default branch, it also adds the milestone when the PR is opened.",
```

Comment:
What's the thought behind not putting incoming PRs into the default milestone when they're opened (but changing milestone if they merge later?)

Reply:
Ooooo, I like that!

Reply:
Reasons against auto-milestoning incoming PR's:
- Mass-migration of 1000 open PR's from one milestone to another every time we make a release (yes, I know, we should not have that many open, but we do)
- Prevents use of `/milestone` as selection criteria for PR's we really care about during code freeze

Reply:
I think my suggestion was a migration on merge, not en masse.

Reply:
Second one I agree with

Reply:
> What's the thought behind not putting incoming PRs into the default milestone when they're opened (but changing milestone if they merge later?)

Yeah, my reasons were same as that what Aaron mentioned.

> I think my suggestion was a migration on merge, not en masse.

IIUC this could still lead to having open PRs with an "old" milestone, right? While the correct/newer milestone would be applied on merge, I think looking at the list of open PRs for a milestone won't provide an accurate representation...

Reply:
Fair point @stevekuznetsov I was reading a bit quickly but yeah, I think it's more meaningful if we allow humans to explicitly declare their intention vs. having the bot forecast, and then have the bot reconcile with reality once things land.